File: code_files_MATLAB/GetVarMatrixParam_Rate.m
% The calculation (aproximation) of the first and second derivative of the likelihood
% function is done by a two side finite differences method
%
% Four methods for the calculation of the covariance matrix were implemented here:
% (1) Using second partial derivatives
% (2) Using first partial derivatives (outer product Matrix)
% (3) Using white covariance matrix (NOT USED)
% (4) Using Newey and West Covariance matrix (NOT USED)
%
% References:
%
% HAMILTON, J., D. (1994). Time Series Analysis. Princeton University Press.
%
% NEWEY, B., WEST, D (1987) ‘A Simple, Positive semidefinite, Heteroskedasticity
% and Autocorrelation Consistent Covariance Matrix’ Econometrica, ed. 55, p. 347-370
%
% WHITE, H. (1984) ‘Asymptotic Theory for Econometricians’ New York:
% Academic Press.

function [V]=GetVarMatrixParam_Rate(fedfund,inflation,outputgap,param)
myDelta=1e-15*abs(param);

% First Derivative calculation
[sumloglik,logLikVec1]=LikelihoodFunc_Rate(fedfund,inflation,outputgap,param);
n=length(logLikVec1); 
s=zeros(numel(param),n);

for i=1:numel(param)
  vec=param;
  vec(i)=param(i)+myDelta(i);
 [sumloglik,logLikVec2]=LikelihoodFunc_Rate(fedfund,inflation,outputgap,vec);

 s(i,:)=(logLikVec2-logLikVec1)/(myDelta(i));
end

OP_Matrix=(s*s')/n;

V=inv(OP_Matrix)/n;




end

File: code_files_R/GetVarMatrixParam_Rate.R
GetVarMatrixParam_Rate <- function(fedfund, inflation, outputgap, param) {
  # --------------------------------------------------------------
  # GetVarMatrixParam_Rate:
  # Computes the variance-covariance matrix of the estimated parameters
  # using the outer product of the gradient (score) method.
  #
  # Methods implemented (only method 2 used):
  # (1) Second partial derivatives (not implemented here)
  # (2) First derivatives (outer product matrix) [used]
  # (3) White covariance matrix (not used)
  # (4) Newey-West covariance matrix (not used)
  # --------------------------------------------------------------
  
  # Small perturbation for finite difference derivative approximation
  myDelta <- 1e-15 * abs(param)
  
  # Baseline likelihood and log-likelihood vector
  base <- LikelihoodFunc_Rate(fedfund, inflation, outputgap, param)
  logLikVec1 <- base$logLik
  n <- length(logLikVec1)
  
  # Initialize score matrix (each column = observation, each row = parameter derivative)
  s <- matrix(0, nrow = length(param), ncol = n)
  
  # Finite difference: compute numerical derivative for each parameter
  for (i in seq_along(param)) {
    vec <- param
    vec[i] <- param[i] + myDelta[i]
    
    out <- LikelihoodFunc_Rate(fedfund, inflation, outputgap, vec)
    logLikVec2 <- out$logLik
    
    s[i, ] <- (logLikVec2 - logLikVec1) / myDelta[i]
  }
  
  # Outer product of gradients
  OP_Matrix <- (s %*% t(s)) / n
  
  # Inverse of OPG (variance-covariance matrix)
  V <- solve(OP_Matrix) / n
  
  return(V)
}
